policy/resource.go

```diff
@@ -136,23 +195,38 @@ func (r Resource) ValidateBucket(bucketName string) error {
 
 // parseResource - parses string to Resource.
 func parseResource(s string) (Resource, error) {
-	if !strings.HasPrefix(s, ResourceARNPrefix) {
-		return Resource{}, Errorf("invalid resource '%v'", s)
+	r := Resource{Type: unknownARN}
+	switch {
+	case strings.HasPrefix(s, ResourceARNPrefix):
+		r.Pattern = strings.TrimPrefix(s, ResourceARNPrefix)
+		r.Type = ResourceARNS3
+
+	case strings.HasPrefix(s, ResourceARNKMSPrefix):
+		r.Pattern = strings.TrimPrefix(s, ResourceARNKMSPrefix)
+		r.Type = ResourceARNKMS
+	default:
+		return r, Errorf("invalid resource '%v'", s)
 	}
 
-	pattern := strings.TrimPrefix(s, ResourceARNPrefix)
-	if strings.HasPrefix(pattern, "/") {
-		return Resource{}, Errorf("invalid resource '%v' - starts with '/' will not match a bucket", s)
+	if strings.HasPrefix(r.Pattern, "/") {
+		return r, Errorf("invalid resource '%v' - starts with '/' will not match a bucket", s)
 	}
 
+	return r, nil
+}
+
+// NewResourceS3 - creates new resource with the default ARN type of AWS S3.
+func NewResourceS3(pattern string) Resource {
```

Comment:
Do we need to break exported API here? Can't we just keep the name. After all the comment says what it does.

Reply:
Sure, I have changed it back.